README.md
# P3 Passionate People Project

## A webapp that allows people to explore jobs based on their passion

### Directory Structure
1. Video for Team 1
2. Interface for Team 2

### Contributor 
(Team 1)

1. Maya Adkins
2. David Hung
3. Caleb Mayerski
4. Gagik Movsisyan
5. Joshua St. Clair

(Team 2)
1. Max Chern 
2. Tremaine Eto 
3. Roy Xia


team2/CHANGELOG
2016-05-31 Tremaine Eto:
	* Implemented World of Work plotting for any given job according to Dr. Scanlan's algorithm
	* Fixed World of Work mapping by rotating 90 degrees

2016-05-31 Roy Xia:
	* Added a mostly empty donors page.
	* Fixed a bug on password reset codes expiring too early
	* Updated password reset UI
	* Added a general 404 page
	* Fixed bug in the browse page partial generation script that was still generating jobs with no education requirement
	* Fixed ordering bug in the browse page partial, it now outputs in SOC code order
	* Implemented dynamic view history retrieval for profile page. No plotting yet.
	* Cleaned up UI for profile page.
	* Implemented random job by World of Work region.

2016-05-30 Tremaine Eto:
	* Set up canvas on top of World of Work to be able to have parameterized x and y values
	* Plot arbitrarily a star on World of Work map for proof of concept
	* Changed "PASSION:" in navibar to a more UCLA-centric (as per client request) gold color

2016-05-30 Max Chern:
	* Fixed Remember Me code to follow MVC

2016-05-29 Roy Xia:
	* Implemented password reset. Not fully robust yet.

2016-05-29 Max Chern:
	* Implemented Remember Me Functionality (Using Cookies/Tokens)

2016-05-28 Max Chern:
	* Reorganized/condensed state data code
	* Implemented Skills Dialog Pie Chart

2016-05-27 Tremaine Eto:
	* Fixed dropdown centering UI bug

2016-05-27 Roy Xia:
	* Refined home page UI when logged in
	* Fixed bug with logout button always being highlighted
	* Reorganized global.css and index.css

2016-05-27 Max Chern:
	* Added definitions to Skills page

2016-05-25 Roy Xia:
	* Refined home page UI

2016-05-25 Tremaine Eto:
	* Implemented dynamic state data for salary page

2016-05-24 Max Chern:
	* Implemented dynamic education page (except state data)
	* Implemented dynamic education page including state data

2016-05-23 Roy Xia:
	* Refactoring of the icons pane
	* Refinement of general site UI
	* Updated scripts to exclude occupations that require less than a Bachelor's degree

2016-05-21 Max Chern:
        * Implemented dynamic skills page

2016-05-20 Roy Xia:
	* Refactored client-side stylesheet and script imports into a partial
	* Implemented responsive web design for the home page

2016-05-20 Max Chern:
        * Implemented LinkedIn Authentication

2016-05-18 Roy Xia:
	* Created script for importing skills data
	* Refactored scripts and refined the script's output to the terminal

2016-05-18 Tremaine Eto:
	* Replaced HTML for ZIP Code with State dropdown for demo during meeting; will now implement with the StateOccupation data Roy imported

2016-05-18 Max Chern:
	* Implemented Facebook Authentication

2016-05-17 Tremaine Eto:
	* Fixed minor data import bug for high salary wage values.
	* Added low average wages and high average wages to Salary Page.

2016-05-17 Roy Xia:
	* Created script for importing occupational interest data.

2016-05-15 Max Chern:
	* Added dynamic Education page information/charts.
	* Fixed the authentication process so that it follows MVC convention.
	* Fixed authentication bugs.
	* Fixed the X button on the sign up pop-up box.

2016-05-13 Tremaine Eto:
	* Minor styling fix: nicely formatted CSS for ZIP code on Salary Page.
	* Fixed grey text bug on Salary Page.

2016-05-13 Roy Xia:
	* Updated the search page so that more information about the job is revealed.
	* Added partial implementation of getting a random career from a World of Work region. Need data on how to plot careers to the World of Work before full implementation.
	* Refactored some utility functions related to formatting into a formatting utility class.
	* Improved search algorithm to allow for keyword searching.
	* Fixed potential SQL injection vulnerabilities.
	* Updated the database schema, which now includes state data, hourly wage data, and low/high wages

2016-05-11 Roy Xia:
	* Implemented search functionality with naive algorithm.
	* Added search bar to browse page and navigation bar.
	* Fixed bug in init_db that was using the wrong account.
	* Fixed styling bug related to fonts.
	* Added UI mockup for browsing by World of Work.

2016-05-11 Tremaine Eto:
	* Added Highcharts JS graph with dynamic, occupation-specific median annual salary data to Salary Page.
	* The reason why Highcharts was chosen was because it reads directly from an HTML table which is actually coded on the HTML page (yet has its display hidden); we have not yet quite figured out how to send the templated Handlebars data on the HTML side to the Javascript side, so this may be something to explore in the future.

2016-05-06 Max Chern:
	* Implemented error messages for bad login/signup attempts.
	* Tweaked index.js; after a bad login/signup attempt, the same login/signup form is displayed.

2016-05-05 Roy Xia:
	* Implemented looking through careers by occupation groups.

2016-05-05 Max Chern:
	* Local Login/Logout Functionality.
	* Navigation bar now presents a logout button if the user is logged in.
	* User can toggle between Sign Up and Log In.
	* Fixed a bug that broke the jQuery functionality on some pages.
	* Moved code to a global css file for global styling.
	* Renamed home page css and js files to index.css and index.js.

2016-05-04 Roy Xia:
	* Created browse page with basic proof of concept.
	* Refactored navigation bar into a partial.
	* Implemented random career functionality.
	* Fixed a bug in the data import script that was producing the opposite value of medianAnnualSalaryOutOfRange.

2016-05-03 Tremaine Eto:
	* Created UI mockup for World of Work page.
	* Added World of Work page to routing mechanism of main.js.
	* Added Profile page to routing mechanism of main.js.

2016-05-03 Roy Xia:
	* Updated Occupation model to support a success and error callback model.
	* Proof of concept for templating career outlook page.
	* Updated routing so that the career video pages and related sub-pages are prefixed with "career" in the resource URL.

2016-05-02 Roy Xia:
	* Added RegionalOccupation table.

team2/models/users.js
var crypto = require('crypto');
var mysql = require('mysql');
var fs = require('fs');
var randomMethods = require('../util/random.js');
var randomstring = require('randomstring');

// Load the database configuration
var config = JSON.parse(fs.readFileSync(__dirname + '/../config/db-config.json', 'utf8'));

// Parameters for PBKDF2. hashSize and saltSize are in bytes.
var hashConfig = { hashSize: 32,
                   saltSize: 16,
                   iterations: 10000 }

module.exports.findById = function(id, next) {
    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM Users WHERE id = '" + id + "';", function(err, rows, fields) {
        if (err === null && rows.length == 1) {
            next(null, rows[0]);
        } else {
            next(new Error('User ' + id + ' does not exist'));
        }
    });

    connection.end();

}

module.exports.findByEmail = function(email, next) {
    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM Users WHERE email = ?", [email], function(err, rows, fields) {
        if (err == null && rows.length == 1) {
            next(null, rows[0]);
        }
        else if (err != null) {
            next(err);
        }
        else {
            next(new Error('User with email not found'));
        }
    });

    connection.end();
}

module.exports.consumeRememberMeToken = function(token, done) {
    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM RememberMeTokens WHERE token = '" + token + "';", function(err, rows, fields) {
        if (err === null && rows.length == 1) {

            connection.query("DELETE FROM RememberMeTokens WHERE token = '" + token + "';", function(err2, rows2, fields2) {
                if (err2 === null) {

                    return done(null, rows[0].id);
                    connection.end();



                } else {
                    return done(err2);
                    connection.end();

                }
            });

        } else {
            return done(new Error('Invalid token'));
            connection.end();
        }
    });
}

module.exports.clearRememberMeToken = function(user, next) {
    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("DELETE FROM RememberMeTokens WHERE id = '" + user.id + "';", function(err, rows, fields) {
        if (err === null) {
            next();
        } else {
            res.writeHead(500);
            res.end('Server error');
        }
    });

    connection.end();
}


//TODO: prevent duplicates?
module.exports.saveRememberMeToken = function(token, id, done) {

    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("INSERT INTO RememberMeTokens VALUES('" + token + "', '" + id + "');", function(err, rows, fields) {
        if (err === null) {
            return done();
        } else {
            return done(err);
        }
    });

}

module.exports.issueRememberMeToken = function (user, done) {
  var token = randomMethods.randomString(64);
  module.exports.saveRememberMeToken(token, user.id, function(err) {
    if (err) { return done(err); }
    return done(null, token);
  });
}

module.exports.signUp = function(req, email, password, done) {

    req.flash('signup', 'yes');

    if (req.body.firstname == '' || req.body.lastname == '' || req.body.verifypassword == '') {
        req.flash('emptyField', 'yes');
        return done(null, false);
    }

    var connection = mysql.createConnection(config);
    connection.connect();

    // First check to see if the email is already taken
    connection.query("SELECT * FROM Users WHERE email = ?", [email], function(err, rows) {
        if (err) {
            connection.end();
            return done(err); //database error
        }
        if (rows.length) {
            connection.end();
            req.flash('emailTaken', 'yes');
            return done(null, false); //email taken
        }
        if (password != req.body.verifypassword) {
            connection.end();
            req.flash('passwordConflict', 'yes');
            return done(null, false);
        }

        // Generate a random salt
        hashPassword(password, function (err, hash, salt) {
            if (err) {
                connection.end();
                /// TODO: Add some user feedback for this error
                return done(null, false);
            }

            connection.beginTransaction(function (err) {
                
                if (err) {
                    connection.end();
                    /// TODO: Add some user feedback for this error
                    return done(null, false);
                }
                
                // Insert into Users...
                connection.query("INSERT INTO Users(firstName, lastName, email) VALUES(?, ?, ?);", [req.body.firstName, req.body.lastName, email], function (err, userInsertResult) {
                    
                    if (err) {
                        connection.end();
                        /// TODO: Add some user feedback for this error
                        return done(null, false);
                    }
                    
                    // ...then into UserPasswords...
                    connection.query("INSERT INTO UserPasswords(hash, salt, id) VALUES(?, ?, LAST_INSERT_ID());", [hash, salt], function (err, userPasswordInsertResult) {
                        if (err) {
                            connection.end();
                            /// TODO: Add some user feedback for this error
                            return done(null, false);
                        }
                        
                        // ...finally, commit the transaction
                        connection.commit(function (err) {
                            connection.end();
                            
                            if (err) {
                                /// TODO: Add some user feedback for this error
                                return done(null, false);
                            }
                            
                            // Return the new user object
                            var newUser = new Object();
                            newUser.firstName = req.body.firstName;
                            newUser.lastName = req.body.lastName;
                            newUser.email = email;
                            newUser.id = userInsertResult.insertId;
                            
                            return done(null, newUser);
                        });
                    });
                });
            });
        });
    });
}


module.exports.login = function(req, email, password, done) {

    req.flash('login', 'yes');

    var connection = mysql.createConnection(config);
    connection.connect();

    // Look up the user associated with the email address
    connection.query("SELECT * FROM Users WHERE email = ?", [email], function(err, userRows) {
        if (err) {
            connection.end();
            return done(err); //database error
        }
        if (!userRows.length) {
            req.flash('incorrectEmail', 'yes');
            connection.end();
            return done(null, false); //no matching email
        }

        // Look up the hashed password and salt for the user
        connection.query("SELECT * FROM UserPasswords WHERE id = ?", [userRows[0].id], function (err, userPasswordRows) {
            if (err) {
                connection.end();
                return done(err);
            }
            if (!userPasswordRows.length) {
                // Not a user that logs in with a password
                connection.end();
                return done(null, false);
            }
            
            // Hash the password that the user is trying to authenticate with
            crypto.pbkdf2(password, userPasswordRows[0].salt, hashConfig.iterations, hashConfig.hashSize, function (err, proposedHash) {
                if (err) {
                    connection.end();
                    return done(err);
                }

                if (proposedHash.toString('hex') !== userPasswordRows[0].hash) {
                    // Wrong password
                    req.flash('incorrectPassword', 'yes');
                    return done(null, false);
                }

                req.flash('success', 'yes');

                // Authentication successful, return the user object
                return done(null, userRows[0]);
            });
        });
    });
}

module.exports.fbAuthenticate = function(accessToken, refreshToken, profile, done) {

    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM FBUsers WHERE fbId = ?", [profile.id], function(err, rows) {
        if (err)
            return done(err);
        if (rows.length) {
            // Join with the Users table to get the object we need to return
            connection.query("SELECT * FROM Users WHERE id = ?", [rows[0].userId], function(err, rows) {
                if (err) {
                    return done(err);
                }
                if (rows.length == 0) {
                    return done(err);
                }

                return done(null, rows[0]);
            });
        } else {
            // Create a new user. Perform this in a transaction because we need to update two tables
            connection.beginTransaction(function (err) {

                if (err) {
                    return done(err);
                }

                // Insert it into Users...
                connection.query("INSERT INTO Users(firstName, lastName, email) VALUES(?, ?, ?);", [profile.name.givenName, profile.name.familyName, profile.emails[0].value], function (err, userInsertResult) {
                    if (err) {
                        return done(err);
                    }

                    // Now insert into FBUsers...
                    connection.query("INSERT INTO FBUsers(fbId, userId) VALUES(?, LAST_INSERT_ID());", [profile.id], function(err, fbInsertResult) {
                        if (err) {
                            return done(err);
                        }

                        // Finally commit
                        connection.commit(function (err) {
                            if (err) {
                                return done(err);
                            }

                            // Return the user object
                            var newUser = new Object();
                            newUser.firstname = profile.name.givenName;
                            newUser.lastname = profile.name.familyName;
                            newUser.email = profile.emails[0].value;
                            newUser.id = userInsertResult.insertId;

                            return done(null, newUser);
                        });
                    });
                });
            });
        }
    });
}

module.exports.liAuthenticate = function(token, tokenSecret, profile, done) {

    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM LIUsers WHERE liId = ?", profile.id, function(err, rows) {
        if (err)
            return done(err);
        if (rows.length) {
            // Join with the Users table to get the object we need to return
            connection.query("SELECT * FROM Users WHERE id = ?", [rows[0].userId], function(err, rows) {
                if (err) {
                    return done(err);
                }
                if (rows.length == 0) {
                    return done(err);
                }

                return done(null, rows[0]);
            });
        } else {
            // Create a new user. Perform this in a transaction because we need to update two tables
            connection.beginTransaction(function (err) {
                if (err) {
                    return done(err);
                }

                // Insert into Users...
                connection.query("INSERT INTO Users(firstName, lastName, email) VALUES(?, ?, ?);", [profile.name.givenName, profile.name.familyName, profile.emails[0].value], function (err, userInsertResult) {

                    if (err) {
                        return done(err);
                    }
                    
                    // Insert into LIUsers...
                    connection.query("INSERT INTO LIUsers(liId, userId) VALUES(?, LAST_INSERT_ID());", [profile.id], function (err, liInsertResult) {

                        if (err) {
                            return done(err);
                        }

                        // Finally commit the transaction
                        connection.commit(function (err) {
                            
                            if (err) {
                                return done(err);
                            }

                            // Return the user object
                            var newUser = new Object();
                            newUser.firstname = profile.name.givenName;
                            newUser.lastname = profile.name.familyName;
                            newUser.email = profile.emails[0].value;
                            newUser.id = userInsertResult.insertId;

                            return done(null, newUser);
                        });
                    });
                });
            });
        }
    });
}

module.exports.issuePasswordResetCode = function(id, next) {
    var connection = mysql.createConnection(config);
    connection.connect();

    // Determine if there's already a pending password reset for this user, if
    // so make sure to update the table rather than insert
    connection.query("SELECT * FROM PendingPasswordReset WHERE id = ?;", [id], function (err, rows, fields) {
        if (err) {
            connection.end();
            next(err);
            return;
        }

        var passwordResetExists = (rows.length != 0);
        // TECH DEBT: We don't handle collisions in the code. The probability
        // of a collision is very low for the expected user count, so currently
        // it's not implemented.
        var code = randomstring.generate(24);
        // Give a 30 minute expiration time on the password reset code
        var expires = new Date(Date.now() + 1000 * 60 * 30);

        if (passwordResetExists) {
            connection.query("UPDATE PendingPasswordReset SET code = ?, expires = ? WHERE id = ?", [code, expires, id], function (err, rows, fields) {
                connection.end();

                if (err) {
                    next(err);
                    return;
                }
                next(null, code);
            });
        }
        else {
            connection.query("INSERT INTO PendingPasswordReset(id, code, expires) VALUES(?, ?, ?)", [id, code, expires], function (err, rows, fields) {
                connection.end();

                if (err) {
                    next(err);
                    return;
                }
                next(null, code);
            });
        }
    });
};

module.exports.lookupPasswordResetCode = function(code, next) {
    var connection = mysql.createConnection(config);
    connection.connect();

    connection.query("SELECT * FROM PendingPasswordReset WHERE code = ?", [code], function(err, rows, fields) {
        if (err) {
            next(err);
            return;
        }

        if (rows.length == 0) {
            next(null, null);
            return;
        }

        next(null, rows[0]);
        return;
    });

    connection.end();
};

module.exports.setPassword = function(id, password, next) {
    hashPassword(password, function(err, hash, salt) {
        if (err) {
            return next(err);
        }

        var connection = mysql.createConnection(config);
        connection.connect();

        // Update the password and simultaneously delete the pending password reset row
        connection.beginTransaction(function(err) {
            if (err) {
                connection.end();
                return next(err);
            }

            // First update the password...
            connection.query("UPDATE UserPasswords SET hash = ?, salt = ? WHERE id = ?", [hash, salt, id], function(err, rows, fields) {
                if (err) {
                    connection.end();
                    return next(err);
                }

                // ...then delete the password reset row
                connection.query("DELETE FROM PendingPasswordReset WHERE id = ?", [id], function(err, rows, fields) {
                    if (err) {
                        connection.end();
                        return next(err);
                    }

                    connection.commit(function(err) {
                        connection.end();

                        return next(err);
                    });
                });
            });
        });
    });
};

// next is a function taking three parameters: err, hash, salt
function hashPassword(password, next) {
    // Generate a random salt
    crypto.randomBytes(hashConfig.saltSize, function (err, bytes) {

        if (err) {
            return next(err);
        }

        var salt = bytes.toString('hex');

        // Generate a hash from the password and salt
        crypto.pbkdf2(password, salt, hashConfig.iterations, hashConfig.hashSize, function (err, bytes) {

            if (err) {
                return next(err);
            }

            var hash = bytes.toString('hex');
            return next(null, hash, salt);
        });
    });
}
